---
title: "Fig_3C"
output: html_notebook
---
```{r}
library(tidyverse)
library(broom)
library(ggtext)
library(microbiomer)
library(dplyr)
library(glue)
```

```{r}
setwd("C:/Users/barak/OneDrive/Noga_Phd/Jellyfish_research/16S_analysis/Microbiome_jellyfish_project/")

# Load phyloseq object WITH CSS
biome_phyloseq<-readRDS("./processed_data/bloom_phyloseq")
sample_data(biome_phyloseq)$Tissue_type <- ifelse(grepl("Bw", sample_data(biome_phyloseq)$Sample_name), "Bell wounded", sample_data(biome_phyloseq)$Tissue_type)

taxa_names(biome_phyloseq) <- paste0("ASV", seq(ntaxa(biome_phyloseq)))

# All samples without controls 
phy_jelly<- biome_phyloseq %>%
                          subset_samples(.,Tissue_type %in% 
                                    c("Bell",
                                      "Tentacles",
                                      "GVC",
                                      "Gonads"))%>%  
                          pres_abund_filter(., pres = 2,abund =0, verbose = TRUE) 

phy_jelly_no_19 <-phy_jelly %>%
                          subset_samples(.,
                                      Year!="2019") %>% 
                          pres_abund_filter(., pres = 2,abund =0, verbose = TRUE) 
```
# Filter otu table to specific number of reads
```{r}

otu_table_filt <- otu_table(phy_jelly_no_19) %>% 
    t() %>%  
    as.data.frame() %>% 
    mutate(Sample = row.names(.)) %>%
    pivot_longer(-Sample) %>% 
    group_by(Sample)  %>%
    mutate(total = sum(value)) %>%
    filter(total > 14000) %>%
    group_by(name) %>%
    mutate(total = sum(value)) %>%
    filter(total > 99) %>%
    ungroup() %>%
    select(-total) %>%
    pivot_wider(names_from = Sample, values_from = value) %>% 
    column_to_rownames(var = "name")

otu_table(phy_jelly_no_19) <- otu_table(otu_table_filt, taxa_are_rows = T)


asv_taxa_meta<-tax_table(phy_jelly_no_19) %>% 
         as.data.frame() %>% 
         rownames_to_column(var = "OTU") %>% 
         mutate(Genus_ASV= glue("{OTU} ({Genus})"))%>% 
         column_to_rownames(var = "OTU") %>% 
         taxtab() 
        
col<-c("Domain", "Phylum" ,"Class" , "Order"  ,"Family" ,"Genus",'Genus_ASV') 
row<-taxa_names(phy_jelly_no_19)

taxa_names(asv_taxa_meta)<-row
colnames(asv_taxa_meta@.Data)<-col

tax_table(phy_jelly_no_19)<-asv_taxa_meta 
```

```{r}
metadata<- sample_data(phy_jelly_no_19) %>% 
           as_tibble(.) %>% 
           select(Sample_name,Month)%>% 
           dplyr::rename(Group=Sample_name)  


shared <- otu_table (phy_jelly_no_19)%>%
          t() %>% 
          as.data.frame () %>% 
          mutate(numOtus=length(colnames(.)),
                 Group= metadata$Group) %>% 
          select(Group, starts_with("ASV")) %>%
          pivot_longer(-Group, names_to="ASV", values_to="count")

taxonomy <- tax_table(phy_jelly_no_19) %>% 
          as.data.frame () %>%
          mutate(ASV= rownames(.))

composite <- inner_join(taxonomy,shared, by="ASV") %>%
  group_by(Group, Genus_ASV) %>%
  summarize(count = sum(count), .groups="drop") %>%
  group_by(Group) %>%
  mutate(rel_abund = 100*(count / sum(count))) %>%
  ungroup() %>%
  select(-count) %>%
  inner_join(., metadata, by="Group")


genus_pool <- composite %>%
  group_by(Month, Genus_ASV) %>%
  summarize(median=median(rel_abund), .groups="drop") %>%
  group_by(Genus_ASV) %>%
  summarize(pool = max(median) < 1,
            median = max(median),
            .groups="drop")


composite_rel_mini <- inner_join(composite, genus_pool, by="Genus_ASV") %>%
  mutate(Genus_ASV = if_else(pool, "Other", as.character(Genus_ASV)),
         Month= factor (Month))%>%
  mutate(Month = fct_relevel(Month, 'Feb', 'Jun', 'Jul')) %>% 
  group_by(Group, Month, Genus_ASV) %>% 
  select(Group, Month, Genus_ASV,rel_abund) %>% 
  filter(Genus_ASV != "Other")
```

In this part I want to preform Kruskal-Wallis test and the pairwise Wilcoxon test for each tissue type in order to conclude what are the significantly taxa in every season./ 
Since I saw different community distribution in different tissues and months, for now I will treat them as separate niches, and will try to characterize each tissue by it self./ 
based on the results I could focus on the main taxa that are relevant for all or focus on the unique taxa of each tissue.

#Kruskal-Wallis and pairwise Wilcoxon tests
```{r}
sig_genera<- composite_rel_mini %>%
  group_by(Genus_ASV) %>%
  nest() %>%
  mutate(experiment_tests = map(.x=data,
                                ~kruskal.test(rel_abund~Month, data=.x,) %>%
                                  tidy())) %>%
  unnest(experiment_tests) %>%
  mutate(p.experiment = p.adjust(p.value, method="BH")) %>%
  select(Genus_ASV, data, p.experiment)%>%
  filter(p.experiment < 0.05)


pairwise_sig_genera <- sig_genera %>%
  mutate(pairwise_tests = map(data, ~pairwise.wilcox.test(x=.x$rel_abund,
                                                          g=.x$Month,
                                                          p.adjust.method = "BH",,
                                                          exact = F) %>%
                          tidy())) %>%
                          unnest(pairwise_tests) %>%
                          filter(p.value < 0.05)


#composite_joined<-composite_rel_mini %>%
 # inner_join(pairwise_sig_genera, by="Genus_ASV")%>%
  #mutate(rel_abund=rel_abund+1/20000) 
  
month_sgnifi_no_19<-ggplot(composite_rel_mini,aes(x=rel_abund,  y=fct_rev(Genus_ASV), color=Month, fill=Month))+
  geom_jitter(position = position_jitterdodge(dodge.width = 0.9,
                                              jitter.width = 0.2),
                                              size=2,
                                              shape=19,
                                              alpha=0.7) +
  stat_summary(fun.data = median_hilow, fun.args = list(conf.int=0.5),
               geom="pointrange",
               position = position_dodge(width=0.9),
               color="black", show.legend = FALSE)+
  coord_trans(x="log10")  +                          # Create the plot with log10 scale This can be good to show the low abundance taxa!
           scale_x_continuous(limits=c(0.01, 100),
                              breaks=c(0,0.1, 1, 10, 100),
                              labels=c(0,0.1, 1, 10, 100)) +
  scale_color_manual(name=NULL,
                   values=c('Feb'='#65DDD9','Jul'="#EADB02","Jun"="#F54441"),
                   limits =c('Feb', 'Jun', 'Jul')) +
  scale_fill_manual(name=NULL,
                   values=c('Feb'='#65DDD9','Jul'="#ddd959","Jun"="#F54441"),
                   limits =c('Feb', 'Jun', 'Jul')) +
  labs(x= "Relative abundance (%)", y=NULL)+
    theme_classic()
```

```{r}
ggsave("./Month_Abundance_signif.pdf", plot = month_sgnifi_no_19, width = 6, height = 6)

```

